### ISSUE TITLE
Prometheus went OOM potentially because of expensive query

### ISSUE THREAD

**Issue by anubshar1** [prometheus/prometheus#14370]
### What did you do?

We had a recent incident in our production prometheus cluster where the memory usage of 2 Prometheus instances suddenly shot up to ~475 GB from the usual ~100 GB, after which the process went OOM and was killed by the kernel. We suspect this was caused by some expensive queries since during the same time, the memory usage of Thanos sidecar also shot up to ~23 GB from the usual ~0.5GB. We have two replicas of Prometheus behind Thanos, running with the exact same config and this happened for both the Prometheus and sidecar instances. The sidecar instances use remote read apis to query from Prometheus.

There has been no change in the trend of the number of queries sent to Thanos/Prometheus so we suspect that there might have been a handful of really heavy queries that caused it. Also there was no unusual increase in the number of timeseries Prometheus has in its head (`prometheus_tsdb_head_series`).

Here's the trend of the memory usage of Prometheus and Sidecar - 

![memory-usage-prometheus](https://github.com/prometheus/prometheus/assets/163815582/4c736a09-a4ce-4740-bf9f-b4b4f9ee372f)

![memory-usage-sidecar](https://github.com/prometheus/prometheus/assets/163815582/7b41df36-2ce2-4a50-b538-ccd4169d87a7)

Prometheus is running with the default limits of 50M for flags `--query.max-samples` and `--storage.remote.read-sample-limit`.  The instances have around ~10M series each in the head.

### What did you see instead? Under which circumstances?

As per our understanding Prometheus memory usage shouldn't increase this much even if very expensive queries are run, mostly because of the fact that it shouldn't bring more than 50M samples per query. Are there any other limits that can be set to keep the memory usage in check? And more importantly, is there anything else we can do to better understand why this spike occurred?

### System information

Linux 3.10.0-1160.118.1.el7.x86_64 x86_64

### Prometheus version

```text
prometheus, version 2.46.0 (branch: HEAD, revision: cbb69e51423565ec40f46e74f4ff2dbb3b7fb4f0)
  build user:       root@42454fc0f41e
  build date:       20230725-12:31:24
  go version:       go1.20.6
  platform:         linux/amd64
  tags:             netgo,builtinassets,stringlabels
```

**Comment by harry671003** [comment]
> As per our understanding Prometheus memory usage shouldn't increase this much even if very expensive queries are run, mostly because of the fact that it shouldn't bring more than 50M samples per query

In Prometheus, the query execution happens in two steps:
- The series and chunks that match the label matchers are selected
- The PromQL expression is evaluated using the selected series and chunks

The `--query.max-samples` limit only protects the second case.
High cardinality queries can still select lots of series and cause Prometheus to go OOM. AFAIK, there is no limit in Prometheus to protect from selecting too many series and chunks.

> Are there any other limits that can be set to keep the memory usage in check?

There isn't any good limit to keep the memory usage in check. Here are a few ideas that might help:
- Ensure that `GOMEMLIMIT` is set to rule out the case of GC not running at the correct time.
- Reduce the number of concurrent queries `--query.max-concurrency`

> And more importantly, is there anything else we can do to better understand why this spike occurred?

You could try introducing continuous profiling to capture the heap profiles. Inspecting the heap profiles before an OOM should give an indication to what caused the high memory usage.

**Comment by beorn7** [comment]
Indeed, Prometheus isn't really designed to have perfect defense against expensive queries. This might change in the future (but would require some fundamental design changes to provide actual guarantees).

I'll convert this issue to a discussion because it contains valuable information how to deal with the problem.
